TF-TRT Improve reshape op converter

tfeher opened the pull request:
This PR improves the Reshape op converter to handle explicit batch and dynamic shape input data.

Since TRT6, IShuffleLayer has an option to accept the shape parameter as an [input tensor](https://docs.nvidia.com/deeplearning/tensorrt/api/c_api/classnvinfer1_1_1_i_shuffle_layer.html#a0ab2ab37e4377c9c317c1d869908efbc). If the shape input tensor corresponds to an input tensor of  the TRTEngineOp, then we need to [specify the input shape values](https://github.com/tfeher/tensorflow/blob/318b5c416db756432125af9d25446bdc3db5ee1f/tensorflow/compiler/tf2tensorrt/utils/trt_engine_utils.cc#L127-L134) for the engine. Additionally we need to specify optimization profile parameters for the shape values. The necessary changes were implemented in TrtShapeOptimizationProfile, TRTEngineOp, trt_engine_utils.

The unit tests have only a single op, and therefore handling shape input values are required to test
reshape with tensor input shape. In normal network it is less likely to have shape input tensors. 

Additionally the handling of scalar input/output shapes are improved.

C++ test for serialization / deserialization of input shapes and shape values added.

Note: TensorRT has a function  [INetworkDefinition::markOutputForShapes](https://docs.nvidia.com/deeplearning/tensorrt/api/c_api/classnvinfer1_1_1_i_network_definition.html#a1f0fff6e9c2469e31ac3be82dd3f0f28) which could enable the output tensor's value to be computed by IExecutionContext::getShapeBinding. This is not used in this PR, but it might be useful for shape inference.

gbaned commented:
@tfeher This PR is in draft, any update on this? Please. Thanks!

gbaned commented:
@tfeher This PR is in draft, any update on this? Please. Thanks!

tfeher commented:
I have updated the PR, it is ready for review.

gbaned commented:
@tfeher  Can you please check @bixia1's comments and keep us posted ? Thanks!

gbaned commented:
@tfeher  Any update on this PR? Please. Thanks!

tfeher commented:
There is an issue with the converter using TRT 7.1.3. I am working on a fix.

gbaned commented:
@tfeher  Any update on this PR? and please resolve conflicts Thanks!

tfeher commented:
The unit test of the converter fails in explicit batch mode with TRT 7.1.3 due to a bug in TRT. Waiting for feedback from the TRT team.

gbaned commented:
@tfeher  Any update on this PR? and please resolve conflicts Thanks!

tfeher reviewed (commented):
Sorry for the delay. I have addressed the first comments, and disable the converter path affected by a bug in TRT. @bixia1 we can continue with the review.

bixia1 commented on tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.h:
```diff
@@ -155,8 +183,12 @@ class TrtShapeOptimizationProfile {
   // Set of input shape vetors that we collect during profile_generation_mode
-  std::unordered_set<std::vector<TensorShape>, VectorTensorShapeHasher>
-      input_shapes_;
+  // Let us consider one of the vectors shape_vec = input_shapes_[0]. The length
+  // of shape_vec is 2 x the number of engine inputs. The first half stores the
+  // shapes of the input tensors. The second half stores shape values if the
+  // tensor is compatible with a TRT shape tensor, otherwise it stores an empty
+  // shape.
+  std::vector<std::vector<TensorShape>> input_shapes_;
```
I can see it is very convenient to append the shape_values to the end of the execution tensor shapes. I have two main questions regarding this:
(1) When selecting a cuda engine: Does OptimizationProfileConfig::IncludesShapes really need to be applied to the shape_values that are actually used as input tensors for reshape?  We also apply the check of OptimizationProfileConfig::IncludesShapes  to the shape values that are NOT used as input tensors for reshape, is that a problem? I tend to think it is a problem, and we will need to fix TrtShapeOptimizationProfile::GetProfileNumber and add test case for this.
(2) Is it possible to delay the copy from device to host of the shape_value in routine AppendShapeValues to TrtShapeOptimizationProfile::ConfigureBuilder? This can avoid unnecessary memcpy.

tfeher replied:
Update: I have modified shape value handling the following way:
- `TrtShapeOptimizationProfile::CollectShapeValues` will store the shape values into the `actual_shape_values_` array. During profile collection, we maintain two vectors: `input_shapes_` and `input_shape_values_`. 
- After profile collection, in `InitProfiles` we concatenate the the vectors from `input_shapes_` and `input_shape_values_`. For both type of inputs we need to perform the same operations when we define the actual profiles, that is the motivation to treat them as a single `std::vector<nvinfer1::Dims>`.

1.b An `OptimizationProfileConfig` contains `2*n_input` values in its `min`, `max`, `opt` members, where `min[n_input + k]` ( 0 <= k < n_input) can describe a shape value input if. If input `k` is not a shape tensor, then the corresponding values will be initialized as `min = opt = max = nvinfer1::Dims{0, {}}`, This is because `CollectShapeValues` zero inits components of `actual_shape_values_` that do not correspond a input shape tensor. Such values will always pass the `IncludesShape` check called by GetProfileNumber.

1.c Once the network is initialized (or engine is loaded from file) we use `SetShapeTensorMask` to query which tensors are actually shape tensors. After this step `CollectShapeValues` will only copy data for inputs that are marked as input shape binding of the engine.

2. Synchronization to ensure that shape values are copied to host: currently it is done in `CollectShapeValues`. This makes the control flow clear. Alternatively we could postpone the sync call to `AppendShapeValues` and `SetInputShapeBinding`.

bixia1 reviewed (commented):
Thanks for the work!  I haven't finished reviewing everything, but I think I should send out the comments I have at this point.

gbaned commented:
@tfeher  Can you please check @bixia1's comments and keep us posted ? Thanks!

gbaned commented:
@tfeher Any update on this PR? Please. Thanks!

tfeher commented on tensorflow/compiler/tf2tensorrt/utils/trt_engine_utils.cc:
```diff
@@ -96,8 +96,9 @@ Status SetTrtEngineInputs(nvinfer1::ICudaEngine* cuda_engine,
-                          const DataVec* input_vec) {
+                          int num_batch,
+                          const std::vector<std::vector<int32>>& shape_values,
+                          OpKernelContext* ctx, const DataVec* input_vec) {
```
This is possible, but we need actually two inputs: the vector of Tensor* and a vector of input names. I had such an implementation earlier, which I rejected to keep `TRTEngineOp::ExecuteTrtEngine` shorter. I can bring this back.

tfeher replied:
I was thinking about the proposed change. I am not convinced if it simplifies things. Let me know if I should go forward with this:

1. For the inputs it could work the following way: 
- we need a helper to construct the vector of tensors from `OpKernelContext` (since the `params_` field is private),
- another helper to map the unit test `DataVec -> gtl::InlinedVector<TensorValue,4>`.

2. The outputs need either to be allocated using `OpKernelContext`, or just reshaped in case of the unit tests. One would need two more helpers that return the vector of tensors: 
- create a helper functions doing the allocations for `OpKernelContext`
- another one for the reshape operations during unit test with `DataVec`.

My preferred solution would be to make the converter tests a subclass of OpTestBase, and run the tests as a TF op test. But that would be outside the scope of this PR.

bixia1 replied:
I think making the converter tests run as OpTestBase is a good idea.

tfeher reviewed (commented):
Thanks @bixia1 for the review and the useful suggestion. There are two issues still open, I am working on them.

tfeher committed:
TF-TRT update reshape op for dynamic shape input

- Add profile management for shape input tensors

- Allow scalars for TRT6 or above

- Improve engine serialization test. Additionally improve shape value
  handling. Adapt to Shape IO API change in TRT 7.1.3.

- Disable ConvertDynamicReshape befor TRT 7.1.3

- Fix Conv3D channel index in implicit batch mode

- Update CombinedNMS scalar value check

DEKHTIARJonathan commented on tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:
```diff
@@ -2250,102 +2280,134 @@ TEST_P(OpConverterTest1, ConvertTranspose) {
 }
 
-TEST_F(OpConverterTest, ConvertReshape) {
+TEST_P(OpConverterTest1, ConvertReshape) {
```
Will need to update with: `OpConverter_FP32_Test`

gbaned commented:
@tfeher  Can you please check @DEKHTIARJonathan's comments and keep us posted ? Thanks!

gbaned commented:
@tfeher Any update on this PR? Please. Thanks!

tensorflowbutler commented:
It has been 15 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

tfeher commented:
This PR depends on #45588, I will update here once #45588 is finalized.

tfeher commented:
@bixia1 this is basically ready for review. But this PR contains the changeset from #45588 which makes it harder to read until #45588 is merged. I will rebase the PR once #45588 is merged.

I have improved the shape value handling as [described here](https://github.com/tensorflow/tensorflow/pull/40545#discussion_r572285311).

tfeher commented:
@bixia1 I have rebased the PR, it is ready for review.

bixia1 commented on tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc:
```diff
@@ -671,14 +672,19 @@ void TRTEngineOp::ComputeAsync(OpKernelContext* ctx,
+    OP_REQUIRES_OK_ASYNC(ctx, cache_res->profiles_.CollectShapeValues(ctx),
+                         *helper);
     if (profile_generation_mode_) {
       // Collecting new shapes for profiles can be only done once. After the
       // shapes are converted to TRT profiles, no shapes can be collected
       // anymore.
-      OP_REQUIRES(ctx, cache_res->profiles_.GetNumProfiles() == 0,
-                  errors::Unimplemented("Cannot collect new shapes when "
-                                        "profiles are already created."));
+      OP_REQUIRES_ASYNC(ctx, cache_res->profiles_.GetNumProfiles() == 0,
+                        errors::Unimplemented("Cannot collect new shapes when "
+                                              "profiles are already created."),
+                        *helper);
       // Just collect the input shape info and return. The shapes are used to
       // generate optimization profiles during engine creation.
       cache_res->profiles_.AddShape(input_concrete_shapes);
```
CollectShapeValues preprares actual_shape_values for AddShape. From this code block, it is possible that AddShape won't be called after CollectShapeValues. It is not straightforward to understand.
Can we combine CollectShapeValues and AddShape and remove actual_shape_values?

bixia1 commented on tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.h:
```diff
@@ -178,16 +211,30 @@ class TrtShapeOptimizationProfile {
-      input_shapes_;
+  std::vector<std::vector<TensorShape>> input_shapes_;
+
+  // Input shape values that we collect during profile_generation_mode. If the
+  // tensor is not compatible with a TRT shape tensor then an empty shape is
+  // stored.
+  std::vector<std::vector<nvinfer1::Dims>> input_shape_values_;
```
Would this be better?
// The tensor values for the inputs with shape-tensor-compatible shapes. 
std::vector<std::vector<absl::optional<nvinfer1::dims>>>> maybe_shape_tensor_values;

bixia1 reviewed (commented):
Thanks for your work! I finish reading most of the code but haven't looked at the tests yet.  I would like to send out my comments and review the test later.

gbaned commented:
@tfeher Can you please check @bixia1's comments and keep us posted ? Thanks!

tfeher reviewed (commented):
Thanks @bixia1 for the review! I have addressed the issues.

tfeher commented:
I have rebased the PR to resolve the merge conflicts.

tfeher commented:
I have noticed a problem with checking the shape value profiles during engine lookup. I am working on a fix.

tfeher commented:
Fixed issue with shape value handling during profile selection, and added the missing datavec.h file. In a subsequent commit I will add one more python test to check shape value handling.

bixia1 reviewed (commented):
Looks like you are still working on the PR. I am sending our the comment I have so far, before your  last commit.

tfeher reviewed (commented):
Thanks @bixia1 for the additional comments. I have corrected the typos. We can continue with the review.

bixia1 commented on tensorflow/compiler/tf2tensorrt/kernels/trt_engine_resource_ops_test.cc:
```diff
@@ -67,42 +80,143 @@ class TRTEngineResourceOpsTest : public OpsTestBase {
+      network = TrtUniquePtrType<nvinfer1::INetworkDefinition>(
+          builder->createNetwork());
+    } else {
+      network = TrtUniquePtrType<nvinfer1::INetworkDefinition>(
+          builder->createNetworkV2(
+              1U << static_cast<int>(
+                  nvinfer1::NetworkDefinitionCreationFlag::kEXPLICIT_BATCH)));
+    }
```
We can only allow this for TRT 6+, right?

bixia1 commented on tensorflow/compiler/tf2tensorrt/kernels/trt_engine_resource_ops_test.cc:
```diff
@@ -223,6 +337,35 @@ TEST_F(TRTEngineResourceOpsTest, Basic) {
+      std::copy_n(vals.data(), vals.size(),
+                  shape_tensor.flat<int32_t>().data());
+      DataVec shape_values{{"one", {}}, {"two", shape_tensor}};
+      // DataVec names are not in CollectShapeValues, used only order matters.
```
Sorry, I don't understand the second part of this comment "used only order matters". Can you explain?
Should this comment be moved to before shape_values?

tfeher reviewed (commented):
Thanks @bixia1 for the review, I have fixed the issues.

bixia1 reviewed (approved):
Approving this so that I can get it to google and test it.

bixia1 commented on tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:
```diff
@@ -763,17 +751,26 @@ Status VerifyShapesMatch(absl::Span<const TRT_TensorOrWeights> inputs,
                                      nvinfer1::Dims dims, Tensor tensor)
-    : shape_(dims), type_(type), tensor_(tensor) {}
+    : shape_(dims), type_(type), tensor_(tensor) {
+  if (dims.d == 0) {
```
do you mean if (dims.nDims == 0) ?

bixia1 commented on tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:
```diff
@@ -2710,12 +2758,47 @@ Status ConvertReshape(OpConverterParams* params) {
+      {{"tensor", TrtInputArg::kTensor}, {"shape", TrtInputArg::kBoth}}));
+  TF_RETURN_IF_ERROR(AllowDataTypes(
+      *params, {DataType::DT_FLOAT, DataType::DT_HALF, DataType::DT_INT32}));
+  const TRT_TensorOrWeights& input_tensor = inputs.at(0);
```
input_tensor is not used.

bixia1 commented on tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc:
```diff
@@ -671,14 +672,18 @@ void TRTEngineOp::ComputeAsync(OpKernelContext* ctx,
   }
 
-  if (!use_implicit_batch_ && has_dynamic_shape_input_) {
+  if (!use_implicit_batch_ && has_dynamic_shape_input_ ||
```
Clang doesn't like this, we need to add parentheses 
s/!use_implicit_batch_ && has_dynamic_shape_input_
/(!use_implicit_batch_ && has_dynamic_shape_input_
)/

bixia1 commented on tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.h:
```diff
@@ -72,28 +77,51 @@ struct OptimizationProfileConfig {
-      profile->setDimensions(name, nvinfer1::OptProfileSelector::kMIN, min[i]);
-      profile->setDimensions(name, nvinfer1::OptProfileSelector::kOPT, opt[i]);
-      profile->setDimensions(name, nvinfer1::OptProfileSelector::kMAX, max[i]);
+      const nvinfer1::Dims input_dim = input->getDimensions();
```
input_dim is not used.

bixia1 commented on tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc:
```diff
@@ -1478,6 +1487,8 @@ class OpConverterTest : public ::testing::Test {
         ProfileStrategy::kImplicitBatchModeCompatible);
     if (!converter_->use_implicit_batch()) {
+      profiles.SetShapeTensorMask(converter_->network());
+      profiles.CollectShapeValues(input_data);
```
Need to handle the return value with TF_RETURN_IF_ERROR

bixia1 commented:
I got three tests failed:
(1) tensorflow/python/compiler/tensorrt:gpu_combined_nms_test. message is like
  File "/build/work/fe0e3a7166e31378da05c1404a4fcd4fedb1/google3/runfiles/google3/third_party/tensorflow/python/compiler/tensorrt/test/tf_trt_integration_test_base.py", line 1107, in _Test
    self.RunTest(run_params)
  File "/build/work/fe0e3a7166e31378da05c1404a4fcd4fedb1/google3/runfiles/google3/third_party/tensorflow/python/compiler/tensorrt/test/tf_trt_integration_test_base.py", line 1030, in RunTest
    GraphState.INFERENCE)
  File "/build/work/fe0e3a7166e31378da05c1404a4fcd4fedb1/google3/runfiles/google3/third_party/tensorflow/python/compiler/tensorrt/test/tf_trt_integration_test_base.py", line 907, in _VerifyGraphDef
    graph_state)
  File "/build/work/fe0e3a7166e31378da05c1404a4fcd4fedb1/google3/runfiles/google3/third_party/tensorflow/python/compiler/tensorrt/test/tf_trt_integration_test_base.py", line 893, in _VerifyGraphDefV2
    [name for name in unexpected_names if name in all_op_names])
AssertionError: ['score_threshold', 'max_output_size_per_class', 'iou_threshold', 'max_total_size'] has length of 4.
(2) tensorflow/python/compiler/tensorrt:gpu_neighboring_engine_test
and tensorflow/python/compiler/tensorrt:gpu_multi_connection_neighbor_engine_test
they both failed because of native segment execution happened,
  (0) Aborted: User disallowed engine native segment execution
	 [[{{node TRTEngineOp_0_1}}]]

bixia1 commented:
tensorflow/compiler/tf2tensorrt:convert_nodes_test_gpu also failed
[test.log](https://github.com/tensorflow/tensorflow/files/6098834/test.log)

tfeher reviewed (commented):
Thanks @bixia1 for the review. I have corrected the smaller issues. I am looking into the python test problems.

The C++ test seem to have a problem with ConvertConv3D. I could not reproduce that wint TRT 7.2.1, but I will have a look with other versions. 

There seem to be an additional issue with shape value handling in TRT6. I will investigate, and disable conversion in the affected versions.

tfeher commented:
- I have disabled ConvertDynamicReshape before TRT 7.1.3, this way the reshape unit test works in all TRT versions that I have tested (5.1.5, 6.0.1, 7.1.3 and 7.2.1). 
- I could not reproduce the Conv3D test with TRT 7.1.3.4.
- I confirm the problems with the two neighbor engine tests. I will continue to investigate those.

bixia1 reviewed (approved):
Per offline conversation, approving this for testing inside google while Tamas is still working on the python tests.

bixia1 commented:
I found the reason for the Conv3D failure. [This line of code ](https://github.com/tensorflow/tensorflow/blob/8cb7f3a462db810d70521fab26eac8e2fed9e2b4/tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc#L3428)is accessing garbage value when c_index == 4, and for implicit batch mode.
In case you need a test to verify it, [here is one](https://github.com/tensorflow/tensorflow/blob/8cb7f3a462db810d70521fab26eac8e2fed9e2b4/tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc#L5018-L5032).
The problem is actually not introduced by this PR, but [this one](https://github.com/tensorflow/tensorflow/pull/46940).

bixia1 reviewed (approved):
Approving this so that I can get it to google and test it while Tamas is still working on one failing test.

tfeher commented on tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.cc:
```diff
@@ -20,6 +20,8 @@ limitations under the License.
 #include "absl/algorithm/container.h"
 #include "tensorflow/compiler/tf2tensorrt/convert/utils.h"
+#include "tensorflow/core/platform/stream_executor.h"
+#include "third_party/gpus/cuda/include/cuda_runtime_api.h"
```
Fixed.

tfeher reviewed (commented):
The NMS Python test error was caused by a change in how scalars are represented in this PR. I have updated the converter, the test passes now.

I have also fixed the two smaller issues.
